### DISCUSSION TITLE
Shape animator modules

### DISCUSSION THREAD

**Discussion by naftalibeder** [rnmapbox/maps#3280] [Q&A] [answered]
@mfazekas I'm iterating on your setup for shape animators to try to get a minimal `MovePointShapeAnimator` example working, and I was wondering what your plan was for handling state.

Inside every subclass of `ShapeAnimatorCommon`, the clock cycle function is `getAnimatedShape(timeIntervalSinceStart)`. To animate smoothly, it needs to know:

- The source coordinate
- The target coordinate
- The time interval since the source and target coordinates were last changed

The functions passed to JS (`create`, `start`, and e.g. `moveTo`) all need to be static, which means it's impossible to do something like:

```swift
@objc
public static func moveTo(tag: NSNumber, coordinate: NSArray) -> Void {
  self.targetCoord = LocationCoordinate2D( // Instance member 'targetCoord' cannot be used on type 'MovePointShapeAnimator'
    latitude: coordinate[1],
    longitude: coordinate[0]
  )
}
```

Given the current setup, with a single manager looping over all animators each tick and just calling `update` on each of them, I'm not sure how to handle this. To work around it would probably take a significant architectural change, unless I'm missing something. Thoughts?

**Comment by mfazekas** [comment, marked as answer]
@naftalibeder so tag is used to look up instance by the manager. (JS side can only address an instance on native side via a tag)


So code could look like this:
```swift
class MovePointAnimatorManager {
@objc
public static func moveTo(tag: NSNumber, coordinate: NSArray) -> Void {
  guard let animator =  ShapeAnimatorManager.shared.get() as? MovePointShapeAnimator else {
     Logger.log(level:.error, "moveTo: Tag: \(tag) wasn't a MovePointAnimator")
  }

  animator.moveTo(LocationCoordinate2D( // Instance member 'targetCoord' cannot be used on type 'MovePointShapeAnimator'
    latitude: coordinate[1],
    longitude: coordinate[0]
  ))
}
}
...
class MovePointAnimator: ShapeAnimatorCommon {
  var from: LocationCoordinate2D;
  var fromTs: TimeInterval = 0.0;
  var to: LocationCoordinate2D;
  var duration: TimeInterval = 10.0;

  func get(timeIntervalSinceStart: TimeInterval) -> LocationCoordinate2D  {
    let progress = (min(duration, timeIntervalSinceStart) - fromTs)/(duration - fromTs);
    LocationCoordinate2D(
      latitude: from.latitude * (1.0 -progress) + to.latitude * progress,
      longitude:  from.longitude * (1.0 - progress) + to.latitude.progress
    )
  }
  
  func getAnimatedShape(timeIntervalSinceStart: TimeInterval) -> GeoJSONObject {
     shapeFromLocation(get(timeIntervalSinceStart: timeIntervalSinceStart))
  }


  func moveTo(coord: LocationCoordinate2D) {
    to = coord;
 
    from = get(progress);
    fromTs = progress;
  }
}
```

I think we can extend ShapeAnimatorCommon a bit, but I think it should support most of the stuf.
